src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java

```diff
@@ -0,0 +1,253 @@
+   * If necessary, assigns the newly found "team project" role to the user,
+   * after validating with a Gen3 authorization service if the user
+   * has been granted access to this team.
+   * 
+   * @param self: reference to "this". Just a workaround to allow the method to be
+   * globally synchronized by making it static.
+   * @param login: the user login
+   * @param request: the request passing through this filter
+   * @param response: response object that can be used to write error status and
+   * error message if needed.
+   * 
+   * @return: Returns true if the "team project" was found and passed validation
+   * and DB update to user roles did not fail.
+   * 
+   * @throws IOException
+   * @throws Exception
+   */
+  private static boolean extractAndValidateTeamProjectRoleAndUpdateUserIfNecessary(
+      TeamProjectBasedAuthorizingFilter self,
+      String login, ServletRequest request,
+      ServletResponse response)  throws IOException, Exception {
+
+    // synchronize on login to avoid race conditions (especially on DB updates) if
+    // this filter receives parallel requests for any reason:
+    synchronized (login.intern()) {
+      // check if a teamproject parameter is found in the request:
+      String teamProjectRole = self.extractTeamProjectFromRequestParameters(request);
+      Set<String> newUserRoles = new HashSet<String>();
+      Set<String> newDefaultRoles = new HashSet<String>(self.defaultRoles);
+
+      // if found, add teamproject as a role in the newUserRoles list:
+      if (teamProjectRole != null && !teamProjectRole.trim().isEmpty()) {
+        // double check if this role has really been granted to the user:
+        if (self.checkGen3Authorization(teamProjectRole, login) == false) {
+          String errorMessage = "User is not authorized to access this team project's data";
+          self.logger.error(errorMessage);
+          WebUtils.toHttp(response).sendError(HttpServletResponse.SC_FORBIDDEN,
+            errorMessage);
+          return false;
+        }
+        // add teamproject role and related system role that
+        // enables read restrictions/permissions based read access configurations:
+        newDefaultRoles.add("read restricted Atlas Users"); // aka reserved system role 15 - TODO: since this is always added...maybe it should be one of the default roles instead...
+        newUserRoles.add(teamProjectRole);
+        self.authorizer.setCurrentTeamProjectRoleForCurrentUser(teamProjectRole, login);
+        self.authorizer.updateUser(login, newDefaultRoles, newUserRoles, true);
+        return true;
+      } else {
+        String errorMessage = "The teamproject is compulsory when on authorizationMode==teamproject configuration";
+        self.logger.error(errorMessage);
+        WebUtils.toHttp(response).sendError(HttpServletResponse.SC_FORBIDDEN,
+          errorMessage);
+        return false;
+      }
+    }
+  }
+
+
```

Comment:
should there be a check for an encrypted channel to sending the authorization?

Reply:
good one. I agree this could be the default behavior (a valid https endpoint - with valid certificate). Not the case right now. 
TODO.

Reply:
Update: I think we should **_not_** enforce it (leaving it to each devops team to decide what's best), but if it is connecting to an external service via HTTPS, then it should fail if the certificate is invalid. 
I tested this and the current code already does the necessary validation (i.e. failing when certificate is self-signed for example). This is how I tested it:
- I've set the property `security.ohdsi.custom.authorization.url=https://reverse-proxy/auth/mapping` where `reverse-proxy` is running on a self-signed certificate. 
   - This resulted in a certificate related error ✅ :
 ```
java.lang.Exception: 
org.springframework.web.client.ResourceAccessException: 
I/O error on POST request for "https://reverse-proxy/auth/mapping": 
PKIX path building failed: 
sun.security.provider.certpath.SunCertPathBuilderException: 
unable to find valid certification path to requested target; 
nested exception is javax.net.ssl.SSLHandshakeException: 
PKIX path building failed: 
sun.security.provider.certpath.SunCertPathBuilderException: 
unable to find valid certification path to requested target
```